js/index.js

```diff
@@ -2,128 +2,798 @@
+      }
+    }).on('reset-layout', function() {
+      $('.-help-column-').each(function(i, element) {
+        $(element).removeClass('-hidden-');
+      });
+      $('.-to-hide-').each(function (i, element) {
+        $(element).addClass('-hidden-');
+      });
+      $(schemaElement).text("");
+    }).on('schema-loaded', function(rawSchema) {
+      var s = location.search.split('schema=')[1];
+      s = s != undefined ? decodeURIComponent(s) : undefined;
+      if (!s || s != rawSchema) {
+        var newUrl = updateQueryStringParameter(location.href, 'schema', rawSchema);
+        // Use this so that it doesn't reload the page, but that also means that you need to manually
+        // load the schema from url
+        window.history.pushState({}, 'AVSC', newUrl);
+        populateFromQuery();
+        eventObj.trigger('update-layout');
+      }
+    }).on('re-instrument', function() {
+      window.instrumented = instrumentObject(window.type, readInput());
+      window.reverseIndexMap = computeReverseIndex(window.instrumented);
+    }).on('update-url', function(data) {
+      var state = {};
+      var newUrl = location.href;
+      for (var key in data) {
+        newUrl = updateQueryStringParameter(newUrl, key, data[key]);
+      }
+      // Use this so that it doesn't reload the page, but that also means that you need to manually
+      // load the schema from url
+      window.history.pushState(state, 'AVSC', newUrl);
 
-  $( function() {
-    var parsedSchema;
-    var encodedErrorElement = $('#encoded-error');
-    var decodedErrorElement = $('#decoded-error');
-    /* Validate schema after each new character. */
-    $('#schema').keyup(function(event) {
-        validateSchema();
     });
 
+    schemaSelect.on('change', function() {
+      loadTemplate(this.value);
+    });
+
+       
+    /* When pasting something into an editable div, it 
+     * pastes all the html styles with it too, which need to be cleaned up.
+     * copied from: 
+     * http://stackoverflow.com/questions/2176861/javascript-get-clipboard-data-on-paste-event-cross-browser */
+
+    $('[contenteditable]').on('paste',function(e) {
+      e.preventDefault();
+      //TODO: Find out why sometimes it triggers 'input-changed' twice.
+      var text = (e.originalEvent || e).clipboardData.getData('text/plain');
+      window.document.execCommand('insertText', false, text);
+      if(e.target.id === 'schema') {
+        if(updateContent(schemaElement)) {
+          eventObj.trigger('schema-changed');
+          generateRandom();
+        };
+      }
+    });
+
+    $('#schema').on('keyup', function() {
+
+      clearTimeout(typingTimer);
+      typingTimer = setTimeout(function () {
+        if(updateContent(schemaElement)) {
+          eventObj.trigger('schema-changed');
+        }
+      }, doneTypingInterval);
+    }).on('keydown', function() {
+      clearTimeout(typingTimer);
+    }).on('click', function(event) {
+      if($(this).hasClass('-placeholder-')) {
```

Comment:
I haven't looked into it but I've heard it's good practice to cache jQuery selectors (`$(this)` here).